src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,179 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    private static class Node<T> {
+        private T item;
+        private Node<T> previous;
+        private Node<T> next;
+
+        public Node(Node<T> previous, T item, Node<T> next) {
+            this.previous = previous;
+            this.item = item;
+            this.next = next;
+        }
+    }
+
     @Override
     public boolean add(T value) {
-        return false;
+        addToTail(value);
+        return true;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index > size || index < 0) {
+            throw new IndexOutOfBoundsException("Index out of Bound");
+        }
+        if (index == 0) {
+            addToHead(value);
+            return;
+        } else if (index == size) {
+            addToTail(value);
+            return;
+        }
+        Node<T> currentNode = findingElement(index);
+        Node<T> node = new Node(currentNode.previous, value, currentNode);
+        currentNode.previous.next = node;
+        currentNode.previous = node;
+        size++;
+    }
+
+    private void addToHead(T value) {
+        Node<T> currentNode = head;
+        Node<T> node = new Node<>(null, value, currentNode);
+        if (currentNode == null) {
+            tail = node;
+        } else {
+            currentNode.previous = node;
+        }
+        head = node;
+        size++;
+    }
+
+    private void addToTail(T value) {
+        Node<T> previousNode = tail;
+        Node<T> node = new Node<T>(previousNode, value, null);
+        if (previousNode == null) {
+            head = node;
+        } else {
+            previousNode.next = node;
+        }
+        tail = node;
+        size++;
+    }
 
+    private Node<T> findingElement(int index) {
+        if (size / 2 >= index) {
+            return searchFromHead(index);
+        } else {
+            return searchFromTail(index);
+        }
+    }
+
+    private Node<T> searchFromTail(int index) {
+        Node<T> currentNode = tail;
+        int localSize = (size - 1) - index;
+        while (localSize > 0) {
+            currentNode = currentNode.previous;
+            localSize--;
+        }
+        return currentNode;
+    }
+
+    private Node<T> searchFromHead(int index) {
+        Node<T> currentNode = head;
+        int localSize = 0;
+        while (localSize != index) {
+            currentNode = currentNode.next;
+            localSize++;
+        }
+        return currentNode;
+    }
+
+    public void checkIndex(int index) {
+        if (index > size - 1 || index < 0) {
+            throw new IndexOutOfBoundsException("Index out of Bound");
+        }
     }
```

Comment:
if you want you can change this method:
```suggestion
    public void checkIndex(int index, int maxSize) {
        if (index > maxSize || index < 0) {
            throw new IndexOutOfBoundsException("Index out of Bound");
        }
    }
```
and then we can use this method in the `add(T value, int index)` too